packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -136,6 +136,86 @@ class AnnotationResult<T> {
+  final Map<int, VoidCallback> _callbacks = <int, VoidCallback>{};
+  static int _nextCallbackId = 0;
+
+  /// Whether the subtree rooted at this layer has any composition callback
+  /// observers.
+  ///
+  /// This only evaluates to true if the subtree rooted at this node has
+  /// observers. For example, it may evaluate to true on a parent node but false
+  /// on a child if the parent has observers but the child does not.
+  ///
+  /// See also:
+  ///
+  ///   * [Layer.addCompositionCallback].
+  bool get subtreeHasCompositionCallbacks => _childrenWithCompositionCallbacks > 0;
+
+  int _childrenWithCompositionCallbacks = 0;
+  void _updateSubtreeCompositionObserverCount(int delta) {
+    assert(delta != 0);
+    _childrenWithCompositionCallbacks += delta;
+    assert(_childrenWithCompositionCallbacks >= 0);
+    if (parent != null) {
+      parent!._updateSubtreeCompositionObserverCount(delta);
+    }
+  }
+
+  void _fireCompositionCallbacks({required bool includeChildren}) {
+    for (final VoidCallback callback in List<VoidCallback>.of(_callbacks.values)) {
+      callback();
+    }
+  }
+
+  bool _debugMutationsLocked = false;
+
+  /// Adds a callback for when the layer tree that this layer is part of gets
+  /// composited, or when it is detached and will not be rendered again.
+  ///
+  /// This callback will fire even if an ancestor layer is added with retained
+  /// rendering, meaning that it will fire even if this layer gets added to the
+  /// scene via some call to [ui.SceneBuilder.addRetained] on one of its ancestor
+  /// layers.
+  ///
+  /// The callback receives a reference to this layer. The recipient must not
+  /// mutate the layer during the scope of the callback, but may traverse the
+  /// tree to find information about the current transform or clip. The layer
+  /// may not be [attached] anymore in this state, but even if it is detached it
+  /// may still have an also detached parent it can visit.
+  ///
+  /// If new callbacks are added or removed within the [callback], the new
+  /// callbacks will fire (or stop firing) on the _next_ compositing event.
+  ///
+  /// Composition callbacks are useful in place of pushing a layer that would
+  /// otherwise try to observe the layer tree without actually affecting
+  /// compositing. For example, a composition callback may be used to observe
+  /// the total transform and clip of the current container layer to determine
+  /// whether a render object drawn into it is visible or not.
+  ///
+  /// Calling the returned callback will remove [callback] from the composition
+  /// callbacks.
+  VoidCallback addCompositionCallback(CompositionCallback callback) {
+    _updateSubtreeCompositionObserverCount(1);
+    final int callbackId = _nextCallbackId += 1;
+    _callbacks[callbackId] = () {
+      assert(() {
+        _debugMutationsLocked = true;
+        return true;
+      }());
+      callback(this is ContainerLayer ? this as ContainerLayer : parent!);
```

Comment:
The callback should probably just always receive a reference to "this" - if the callback wants, it can then check whether it was given a non-ContainerLayer and access the parent.

Reply:
The callback was written to receive a `ContainerLayer`, which has all the interesting methods for traversal on it.

In the layer tree, leaf `Layer` objects always have a parent at compositing time - in fact, this method is called from `buildScene`, which only exists on `ContainerLayer`. If there was no container layer parent, then this method would never get called.

Reply:
If the callback was typed to receive a `Layer`, for example, and a caller wanted to start reviewing the clip bounds applied, it woud have to do something like `ContainerLayer container = layer is ContainerLayer ? layer : layer.parent!`.

Reply:
I'll ggo ahead and change this now - I suppose if the implementation ever shifted or some new more interesting method for this is added on `Layer` it'd be better to not have to refactor this API later.

Reply:
Also moved `describeClipBounds` to layer.